### ISSUE TITLE
DPL-961 Bioscan - Research - RT #790511, 790510 New 'used up' state for LILYS-96 Stock plate purpose

### ISSUE THREAD

**Issue by SujitDey2022** [sanger/limber#1435]
**User story**
As R&D (Abdul T/Scott T) we would like the ability to flag (with a new state label) wells on LILYS-96 Stock and/or  LBSN-96 Lysate plates within the BIOSCAN non-destructive extraction pipeline in order to identify wells that are not suitable for further transfers. e.g. the insect has been lost from the well, or the well is otherwise contaminated or unusable. Flagging the well in such a way should not affect any library prep in progress, just future uses.

**Who are the primary contacts for this story**
Scott T (R&D)

Abdul T (R&D)

Emma B (R&D)

**Who is the nominated tester for UAT**
Scott T (R&D)

Abdul T (R&D)

Emma B (R&D)

**Acceptance criteria**
To be considered successful the solution must allow:

From RT https://rt.sanger.ac.uk/Ticket/Display.html?id=790511

Allow the following actions to be carried out on the LILYS-96 Stock plate purpose:

- *new state for individual sample wells (name TBC e.g. 'used up' or 'exhausted')
  - Mark one or more wells with the new state
  - This should prevent further requests for work being generated on those wells
  - This functionality should be available on the LILYS-96 Stock plate purpose after Beckman LILYS-96 Stock Preparation and Beckman LILYS-96 Stock => LBSN-96 Lysate steps
- If sample is selected (as part of a whole plate transfer), the child well should be marked in some way to indicate no sample exists in this well
- Behaviour detailed further in ‘Research User Story : New cancel/fail functionality’

From RT https://rt.sanger.ac.uk/Ticket/Display.html?id=790510

- Apply only to selected plate purposes
- Allow selection of individual samples on identified plate purposes
- Allow progression of all existing child samples of the selected sample
- Prevent creation of future child samples/submissions from a given plate purpose
  - If sample is selected (as part of a whole plate transfer), the child well should be marked in some way to indicate no sample exists in this well
  - This sample should be marked as unavailable in sample selection/cherry-picking criteria
- Indicate the sample failure to the user in the LIMS UI
  - Ideally this would be visually distinct from ‘failed’ or ‘cancelled’ samples which use a red colouring.

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

References
This story has a non-blocking relationship with:
- https://github.com/sanger/limber/issues/1434

**Additional context**

From RT https://rt.sanger.ac.uk/Ticket/Display.html?id=790510

Extending the cancel functionality to sample level may achieve the desired result.  

**BIOSCAN use case**

In the LILYS-96 Stock plate to  LBSN-96 Lysate plate transfer, lysed material containing specimen DNA is transferred from the LILYS plate (which contains insect specimens) to the Lysate plate (which will proceed to library prep). 

During this transfer, there is a risk that the specimen will be lost (dropped, aspirated into the tip, stuck to outside of tip, etc.).

The lysed material which has been transferred to the Lysate plate can proceed to library prep as normal (as it contains DNA), but as the insect specimen is no longer present in the LILYS plate, no further extraction is possible, so new child samples should be prevented from being created. 

**RVI cherry-picking use case**

The RVI project will utilise multiple library preparation workflows to identify the detailed composition of the stock sample.

Samples may be identified as requiring multiple library prep types at the outset, or the outcome of one library prep, may trigger the generation of one or more different library preps.

The number of library preps required from a stock sample cannot be determined at the outset, each library prep will have different input requirements, and the volume of each sample varies greatly and is not directly recorded.  

The RVI sample selection process that feeds into cherry-picking needs some mechanism to indicate that a stock sample cannot generate more child samples, without failing the libraries that have already been created and sequenced.

**Comment by SujitDey2022** [comment]
@andrewsparkes created the common Research story for https://rt.sanger.ac.uk/Ticket/Display.html?id=790511 and https://rt.sanger.ac.uk/Ticket/Display.html?id=790510

I have not created the actual implementation story as I believe the research story may lead to multiple small stories.

Please could you help review and update the story as applicable.

**Comment by TWJW-SANGER** [comment]
@andrewsparkes Quick check - is this story still required with your new fix.

**Comment by andrewsparkes** [comment]
Yes this is still required.

Currently well state is 'calculated' based on the state of the transfer into the well for intermediate plates, whereas input plates have well state set to a default or 'unknown'. And 'Sample' is abstract and doesn't have state. So not straightforward.

**Comment by KatyTaylor** [comment]
Just writing up my thoughts on this before the Christmas break - having spent about a day looking at it so far.

**Potential LIMS touch points - may not need all:**
1. Bioscan library prep
    - Prevent submission (Sequencescape submissions)
    - Prevent transfer (Limber pipeline)
    - Visualise well state (Limber plate page)
2. RVI cherrypicking
    - Prevent submission (Sequencescape submissions)
    - Prevent transfer (Sequencescape cherrypicking pipeline)
    - Visualise well state (Sequencescape plate page, and/or cherrypicking pipeline)

**Potential UX ideas:**
1. Bioscan library prep
    - Integrate into existing 'fail wells' user interface, allowing a reason to be specified, and colouring the wells differently
        - Note appears saying it will not affect downstream wells, unlike normal failing.
    - Alternatively, well failing interface could have a checkbox that says 'Remove from downstream wells' or similar, which defaults to checked, but could be unchecked in this case.
2. RVI cherrypicking
    - Have not yet thought about UX - using Limber may not be appropriate here as this work is done in SS.

**Possible implementation approaches:**
1. _Use a new state, 'used up' or 'exhausted', in a similar manner to existing states._ Existing well state is 'calculated' on demand read from either the Transfer Requests coming into the well (standard), or the Requests coming out of the well (for input plates, in progress implementation by Andrew). Problem here is that the 'failure' is not related to a transfer request or a request. In both use cases, the well acquires the 'used up' state after a successful request / transfer out of the well. It wouldn't be correct to record the state on that last successful transfer - it is a property of the well itself. It makes more sense to record it against the well, whether directly or on a related object.
2. _Store state against well / receptacle._ This makes more sense than the above. I need to look into whether this already happens in any way. One risk here is that we are creating a new way of doing an existing thing, and therefore adding complexity to the product. It might make sense to refactor all well state functionality to use this, but that is a bigger piece of work.
3. _Use one of the existing ways of storing 'volume'._ Used up / exhausted wells would have a volume of 0. There are various features around volume already. I did some research and recorded my (not yet comprehensive) findings in [this diagram](https://lucid.app/lucidchart/ef7cb0c3-cba7-42e6-8c4a-79aa8bbff9fb/edit?viewport_loc=-1032%2C-360%2C6976%2C3800%2C0_0&invitationId=inv_c5f9b452-b992-4758-97ed-3f4dcf49f524). I think this is promising, because it taps into existing functionality rather than creating additional complexity. The risk is whether the volume data is trustworthy or not, as it is stored in many different ways and we know is fudged in some cases (e.g. 'fake cherrypicking').
4. _Use a joined up solution with long read for volume tracking._ There is some research currently going into volume tracking for long read - to be stored in the MLWH for external visibility.

**Comment by KatyTaylor** [comment]
Thinking through one option for implementation:

**Where to store the data?**

- Add new ‘unusable’ (boolean) and ‘unusable_reason’ (text) fields, on Receptacle or equivalent metadata table.

**How to set the data?**

- Set both fields automatically when a receptacle is failed or cancelled. Only makes sense on receptacles that might be part of a submission, but should do no harm on others.

- Could consider adding an option to mark the stock receptacle as exhausted, when failing a downstream well, if there is a user requirement for it (could probably be an enhancement if desired).

- Provide a way to set them manually - new button on Limber plate page, ‘Mark wells as unusable’ or similar, uses same style of interface as failing wells. Also provide similar UI for tubes.

- (To be thought through further) Set both fields automatically after cherrypicking, if purposes are in a configured list (initially just for RVI).

- N.B. Setting these fields has no side effects, i.e. existing requests and downstream wells are unaffected.

**How to view the data?**

The two fields should be visible:
- On a Sequencescape Receptacle page, as text
- On a Limber plate page, as a coloured well (different colour to ‘failed’)
- On a Limber tube page, as a different colour (look into whether tubes can be ‘failed’ and how this looks)

**How to use the data?**

When creating a submission, raise an exception if any of the receptacles are ‘unusable’
- Make sure to handle this in the SS interface, and in Limber when creating submissions through the API, and anywhere else where submissions can be created (if there are any?)

**Questions / concerns**

- Should we use a ‘state’ instead of a flag? Would provide the opportunity to change how well ‘state’ from using requests to using this same field. But might be better to keep separate.

**To do**

- Make wireframes to show intended UX.
- Make stories (most small I think)
    - Creation of new fields in db
    - Submission block
    - UI for manually setting them
    - Automatic setting on failure / cancellation of wells
    - Automatic setting on cherrypicking
    - Add to SS Receptacle page
    - Add to Limber plate page
    - Add to Limber tube page

**Comment by KatyTaylor** [comment]
I'm going to close this one off today, so here's my final write up:

### TL;DR

- The 'exhausting wells' feature could be useful, as a distinct feature from failing wells - where failing wouldn't make sense because nothing has gone wrong with the well, but there is no material left so future work from that well should be avoided.
- I think the design in the comment above works for the use cases I have seen.
- However, this work is not high priority for now, as it is not needed any more for Bioscan (due to low error rate), and not urgent for RVI (due to low throughput), so it can go on the back-burner.

---------

### Details

The use cases (including those for [DPL-960](https://github.com/sanger/limber/issues/1434)) were written up here - https://docs.google.com/document/d/1xknIM1Rh5RMkPjZUMVXBwZS2JHGwWCGzif_stcn6tyE/edit?usp=sharing - this is more comprehensive and correct than this story's description. Short version of the 'exhausting wells' use cases below (use case numbers matching doc), in case something happens to the doc.:

> **Bioscan Use case 4:** 
> Marking a stock well as 'exhausted' and failed, when an insect is lost after addition of lysis buffer (first robot step) - meaning the current run has failed.
> 
> **Bioscan Use case 5:** 
> Marking a stock well as 'exhausted', when an insect is lost after lysate transfer (second robot step) - meaning the current run is OK to proceed, but future runs are not possible.
> 
> **RVI Scenario 1:** 
> User manually marks one or more wells of LDW-96 Stock plate as exhausted as they were supplied with no volume and we would like to prevent sample selection in cherrypicking. 
> - Setting a sample to exhausted here should not prevent the LDW-96 Stock to LSW-96 Stock transfer step, but not carry forward sample data to the LSW-96 Stock plate
> - It should also not prevent the creation of a LSW-96 Stock plate from a LDW-96 Stock plate containing an exhausted sample
> 
> **RVI Scenario 2:** 
> User manually marks one or more wells of LSW-96 Stock plate as exhausted as they were lost during the LDW-96 Stock to LSW-96 Stock transfer step.
> - The LDW-96 Stock plate may have been empty before the transfer, but it is possible we can only confirm that at this stage. 
> 
> **RVI Scenario 3:** 
> System automatically marks all wells cherrypicked from source plate into RVI Cherrypick as exhausted as part of cherrypick process.
> - Supported RVI Cherrypick inputs LDW-96 Stock, LSW-96 Stock, RVI Cherrypick, Stock Plate.
> 
> **RVI Scenario 4:** 
> User manually marks one or more wells of the RVI Cherrypick plate as exhausted after the cherrypick process is complete in the case that sample was lost during this process.
> - This should not prevent transfer of other samples from RVI Cherrypick to RVI RT
> - This should prevent or flag to user that this sample is not available for selection for future cherrypicking
> 
> **RVI Scenario 5:** 
> User manually marks one or more wells of RVI Cherrypick plate as exhausted after the RVI Cherrypick to RVI RT stamp as all remaining sample has been taken.
> - This should not prevent transfer of other samples from RVI Cherrypick to RVI RT
> - This should prevent or flag to user that this sample is not available for selection for future cherrypicking


The Bioscan use cases were crossed off that list because:
- Originally, the customer thought there would be a higher incidence of losing insects from wells
- Therefore they wanted fine-grained reporting of failure reasons
- However, the actual failure rate is v low (19 in several thousand) so this reporting is no longer necessary

The first two RVI scenarios were crossed off the list due to communication from Anna G in email thread with subject "Sample selection for RVI", and subsequent meeting on 26/01/24:
- They can use existing well failing functionality on the shallow well plates (they already have, for the 5k that have been stamped already), and manually check (in Limber) any plates selected for cherrypicking, to see if any wells are failed.
It's not really possible to see if there's any material present in the deep well plates, so any issues with the deep well plates will be marked against the shallow well plates.

The remaining three RVI use cases would still be useful. Even with low throughput, keeping track manually of which wells have been exhausted will be difficult and error prone, because there would likely be a long period between when a plate was first picked from, and when it was next picked from.

However, the throughput is likely to be only one plate a week (concentrating on 'current' delivered samples, rather than historical samples) until the lab is moved to Sulston. In addition, the GSLP team are moving to a new LIMS at some point. For these reasons, the priority of building this feature for them is fairly low.

If we did build the feature for RVI, I think the relevant stories would be as follows (see previous comment for design details):
1. Add new ‘unusable’ (boolean) and ‘unusable_reason’ (text / drop down) fields, on Receptacle or equivalent metadata table.
2. Add 'Exhaust wells' feature in Limber UI, similar to 'Fail wells' feature, so wells can be manually exhausted.
3. Automatically mark source wells as exhausted after cherrypicking, for specific RVI plate purposes.
4. (maybe) Create a report to help customer see either 
a) from a list of samples, which stock / cherrypick plate their sample is in, and not exhausted, or
b) which wells are exhausted / not, from a list of plate barcodes
This potential story should be discussed with Ewan, as he is the one doing the sample selection and Anna wasn't sure exactly what his process would be or what data he would be looking at to decide.
If this story is not done, the alternative is that he sends a list for cherrypicking to GSLP (Anna), and they manually check each plate in Limber to see if there are any wells failed or exhausted, before sending the list on to the SSRs (pre-formatted to paste into the cherrypicking submission box).

N.B. Anna is checking with Adrienne that the above discussion about RVI priority is correct.

**Comment by TWJW-SANGER** [comment]
Closing.
From Scott T:
Faculty are happy to drop the request for an ‘exhausted’ well state as they will not re-extract from LILYS plates. They have tested re-extraction since the initial request was made and results were poor.
